src/PostgREST/Cache/Sieve.hs

```diff
@@ -0,0 +1,218 @@
+{-|
+Module      : PostgREST.Cache.Sieve
+Description : PostgREST cache implementation based on Sieve algorithm.
+
+This module provides implementation of a mutable cache on Sieve algorithm.
+-}
+{-# LANGUAGE DataKinds       #-}
+{-# LANGUAGE GADTs           #-}
+{-# LANGUAGE LambdaCase      #-}
+{-# LANGUAGE NamedFieldPuns  #-}
+{-# LANGUAGE PolyKinds       #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE RecursiveDo     #-}
+{-# LANGUAGE StrictData      #-}
+{-# LANGUAGE TupleSections   #-}
+
+module PostgREST.Cache.Sieve (
+      Cache
+    , CacheConfig (..)
+    , Discard (..)
+    , alwaysValid
+    , cache
+    , cacheIO
+    , cached
+)
+where
+
+import           Control.Concurrent.STM
+import           Control.Monad.Extra    (whileM)
+import           Data.Some
+import qualified Focus                  as F
+import           Protolude              hiding (elem, head)
+import qualified StmHamt.SizedHamt      as SH
+
+data ListNode k v (b :: Bool) = ListNode {
+        nextPtr        :: NodePtr k v,
+        prevNextPtrPtr :: NodePtrPtr k v,
+        elem           :: NodeElem k v b
+    }
+
+data NodeElem :: Type -> Type -> Bool -> Type where
+    Head :: {
+            entries          :: SH.SizedHamt (HamtEntry k v),
+            finger           :: NodePtrPtr k v
+        } -> NodeElem k v False
+    Entry :: Hashable k => {
+            visited :: TVar Bool,
+            ekey :: k,
+            entryValue :: v
+        } -> NodeElem k v True
+
+type HamtEntry k v = ListNode k v True
+type AnyNode k v = Some (ListNode k v)
+type NodePtr k v = TVar (AnyNode k v)
+type NodePtrPtr k v = TVar (NodePtr k v)
+
+data Discard m v = Refresh (m ()) | Invalid (m v)
+
+data Cache m k v = (MonadIO m, Hashable k) => Cache (ListNode k v False) (CacheConfig m k v)
+
+data CacheConfig m k v = CacheConfig {
+        maxSize          :: STM Int,
+        load             :: k -> m v,
+        requestListener  :: Bool -> m (),
+        evictionListener :: k -> v -> m (),
+        validator        :: m (k -> v -> Maybe (Discard m v))
+}
+
+alwaysValid :: Applicative m => m (k -> v -> Maybe (Discard m v))
+alwaysValid = pure (const . const Nothing)
+
+cacheIO :: (MonadIO m, Hashable k) => CacheConfig m k v -> IO (Cache m k v)
+cacheIO = atomically . cache
+
+cache :: (MonadIO m, Hashable k) => CacheConfig m k v -> STM (Cache m k v)
+cache cacheConfig = mdo
+    tail <- newTVar (Some head)
+    entries <- SH.new
+    finger <- newTVar tail
```

Comment:
slightly pedantic, this should be `hand` right?  Just understood the algorithm from the blog post and they use the term `hand`.